crates/bevy_ecs/src/storage/table/mod.rs

```diff
@@ -0,0 +1,860 @@
+            col.changed_by
+                .initialize_unchecked(len, UnsafeCell::new(Location::caller()));
+        }
+        TableRow::from_usize(len)
+    }
+
+    /// Gets the number of entities currently being stored in the table.
+    #[inline]
+    pub fn entity_count(&self) -> usize {
+        self.entities.len()
+    }
+
+    /// Get the drop function for some component that is stored in this table.
+    #[inline]
+    pub fn get_drop_for(&self, component_id: ComponentId) -> Option<unsafe fn(OwningPtr<'_>)> {
+        self.get_column(component_id)?.data.drop
+    }
+
+    /// Gets the number of components being stored in the table.
+    #[inline]
+    pub fn component_count(&self) -> usize {
+        self.columns.len()
+    }
+
+    /// Gets the maximum number of entities the table can currently store
+    /// without reallocating the underlying memory.
+    #[inline]
+    pub fn entity_capacity(&self) -> usize {
+        self.entities.capacity()
+    }
+
+    /// Checks if the [`Table`] is empty or not.
+    ///
+    /// Returns `true` if the table contains no entities, `false` otherwise.
+    #[inline]
+    pub fn is_empty(&self) -> bool {
+        self.entities.is_empty()
+    }
+
+    /// Call [`Tick::check_tick`] on all of the ticks in the [`Table`]
+    pub(crate) fn check_change_ticks(&mut self, change_tick: Tick) {
+        let len = self.entity_count();
+        for col in self.columns.values_mut() {
+            // SAFETY: `len` is the actual length of the column
+            unsafe { col.check_change_ticks(len, change_tick) };
+        }
+    }
+
+    /// Iterates over the [`ThinColumn`]s of the [`Table`].
+    pub fn iter_columns(&self) -> impl Iterator<Item = &ThinColumn> {
+        self.columns.values()
+    }
+
+    /// Clears all of the stored components in the [`Table`].
+    pub(crate) fn clear(&mut self) {
+        let len = self.entity_count();
+        for column in self.columns.values_mut() {
+            // SAFETY: we defer `self.entities.clear()` until after clearing the columns,
+            // so `self.len()` should match the columns' len
+            unsafe { column.clear(len) };
+        }
+        self.entities.clear();
+    }
+
+    /// Moves component data out of the [`Table`].
+    ///
+    /// This function leaves the underlying memory unchanged, but the component behind
+    /// returned pointer is semantically owned by the caller and will not be dropped in its original location.
+    /// Caller is responsible to drop component data behind returned pointer.
+    ///
+    /// # Safety
+    /// - this table must hold the component matching `component_id`
+    /// - `row` must be in bounds
+    pub(crate) unsafe fn take_component(
+        &mut self,
+        component_id: ComponentId,
+        row: TableRow,
+    ) -> OwningPtr<'_> {
+        self.get_column_mut(component_id)
+            .debug_checked_unwrap()
+            .data
+            .get_unchecked_mut(row.as_usize())
```

Comment:
I don't see anything here that results in the component being forgotten, this seems to just grab a mutable reference and return it

Okay I found the use site and it says you need to remove the row afterwards, I wonder if Safety should mention that here too...